File: src/engines/numerology/services/aggregator.py
"""
Logic to blend multiple elemental vectors based on specific weights.
"""
from ..constants import parameters

def mix_profiles(life_path_vec, expression_vec, soul_urge_vec):
    """
    Weighted mixing of the three vectors.
    """
    
    # Initialize result vector
    final_vector = {"air": 0.0, "water": 0.0, "fire": 0.0, "earth": 0.0, "spirit": 0.0}
    
    # Helper to add weighted values
    def apply_weight(vector, weight):
        if not vector: return
        for element, value in vector.items():
            final_vector[element] += (value * weight)

    # Apply specific weights
    apply_weight(life_path_vec, parameters.WEIGHTS['life_path'])
    apply_weight(expression_vec, parameters.WEIGHTS['expression'])
    apply_weight(soul_urge_vec, parameters.WEIGHTS['soul_urge'])
    
    # Rounding for cleanliness
    for key in final_vector:
        final_vector[key] = round(final_vector[key], 4)
        
    return final_vector

File: src/server/engines/numerology/aggregator.ts
import type { ElementVector } from "@/types/personalization";
import { WEIGHTS } from "./constants";

const ELEMENTS = ["air", "water", "fire", "earth", "spirit"] as const;

/**
 * Weighted blend of life path, expression, and soul urge vectors.
 * LP 50%, Expression 30%, Soul Urge 20%.
 * If suVec is null, reweight LP and Expression proportionally.
 */
export function mixProfiles(
  lpVec: ElementVector,
  expVec: ElementVector,
  suVec: ElementVector | null,
): ElementVector {
  let lpWeight: number;
  let expWeight: number;
  let suWeight: number;

  if (suVec === null) {
    const totalWithout = WEIGHTS.life_path + WEIGHTS.expression;
    lpWeight = WEIGHTS.life_path / totalWithout;
    expWeight = WEIGHTS.expression / totalWithout;
    suWeight = 0;
  } else {
    lpWeight = WEIGHTS.life_path;
    expWeight = WEIGHTS.expression;
    suWeight = WEIGHTS.soul_urge;
  }

  const result: ElementVector = { air: 0, water: 0, fire: 0, earth: 0, spirit: 0 };

  for (const el of ELEMENTS) {
    const blended =
      lpVec[el] * lpWeight +
      expVec[el] * expWeight +
      (suVec !== null ? suVec[el] * suWeight : 0);
    result[el] = Math.round(blended * 10000) / 10000;
  }

  return result;
}
